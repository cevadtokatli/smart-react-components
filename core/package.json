{
  "name": "@smart-react-components/core",
  "version": "2.0.0-beta-0",
  "license": "MIT",
  "main": "index.js",
  "description": "SRC core includes the theme, Styled properties, and Element components.",
  "keywords": [
    "javascript",
    "typescript",
    "react",
    "styled-components"
  ],
  "author": {
    "name": "Cevad Tokatlı",
    "email": "cevadtokatli@hotmail.com"
  },
  "contributors": [
    {
      "name": "Cemil Tokatlı",
      "email": "cemil@tokatli.me"
    }
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/cevadtokatli/smart-react-components/tree/v2/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "yarn clear && tsc --outDir lib && yarn copy",
    "clear": "rm -rf lib",
    "copy": "cp package.json lib/package.json && cp README.md lib/README.md && cp ../LICENSE lib/LICENSE",
    "dev": "tsc -w --outDir lib",
    "lint": "eslint src",
    "release": "../scripts/release.sh core",
    "test": "jest"
  },
  "devDependencies": {
    "@types/color": "3.0.3",
    "@types/enzyme": "3.10.12",
    "@types/jest": "29.0.1",
    "@types/react": "18.0.19",
    "@types/react-dom": "18.0.6",
    "@types/styled-components": "5.1.26",
    "@wojtekmaj/enzyme-adapter-react-17": "0.6.7",
    "enzyme": "3.11.0",
    "eslint": "8.0.1",
    "jest": "29.0.3",
    "jsdom": "20.0.0",
    "jsdom-global": "3.0.2",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "styled-components": "5.3.5",
    "ts-jest": "29.0.0",
    "typescript": "4.8.3"
  },
  "dependencies": {
    "color": "4.2.3"
  },
  "peerDependencies": {
    "react": "^18.2.0",
    "styled-components": "^5.3.5"
  }
}
