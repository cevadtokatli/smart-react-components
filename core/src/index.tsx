export * from './theme'
export { default as createTheme } from './theme'
export * from './styled-props'
export * from './Element'
