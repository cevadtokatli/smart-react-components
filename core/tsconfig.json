{
  "extends": "../tsconfig.json",
  "include": ["src/**/*.tsx"],
  "exclude": ["node_modules", "src/**/*.test.tsx"]
}
