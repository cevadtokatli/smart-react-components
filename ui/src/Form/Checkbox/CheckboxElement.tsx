import Div from '@smart-react-components/core/Element/Div'
import { StyledProps } from '@smart-react-components/core/styled-props'
import { PaletteProp, ResponsiveProp, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import styled from 'styled-components'

const Container = styled.div(({ theme }) => `
  border: solid 2px ${theme.$.color.dynamic.accent};
  box-sizing: border-box;
  display: flex;
  height: 100%;
  justify-content: center;
  position: relative;
`)

const Rectangle = styled.div`
  position: absolute;
  left: -2px;
  height: calc(100% + 4px);
  top: -2px;
  transition: background 200ms 0s ease-in-out;
  width: calc(100% + 4px);
`

const Svg = styled.svg.attrs({
  children: <path d="M3.5,17.2l7.6,7.6L28.5,7.3" />,
  viewBox: '0 0 32 32',
})(({ theme }) => `
  flex: 0 0 auto;
  position: relative;

  > path {
    fill: transparent;
    stroke: ${theme.$.color.dynamic.background};
    stroke-width: 4px;
    stroke-dasharray: 36px;
    stroke-dashoffset: 36px;
    transition: stroke-dashoffset 200ms 0s ease-in-out;
  }
`)

interface Props extends
  StyledProps,
  ResponsiveProp<'checkboxSize', SizeProp> {
  isChecked: boolean
  isOutline: boolean
  isSoft: boolean
  palette: PaletteProp
  shape: ShapeProp
}

export default styled(Div).attrs<Props>(({ children, isOutline, isSoft, palette, shape }) => ({
  children: (
    <>
      { children && children }
      <Container>
        <Rectangle />
        <Svg />
      </Container>
    </>
  ),
  getCheckboxSize: (v, t) => `
    height: ${t.$.size.form.checkbox[v]};
    width: ${t.$.size.form.checkbox[v]};
  `,
}))<Props>(({ theme, isChecked, isOutline, isSoft, palette, shape }: Props) => `
  position: relative;

  ${Container} {
    border-radius: ${theme.$.radius.form.checkbox[shape]};
  
    ${!isOutline
      ? `
        background: ${theme.$.color.dynamic.accent};
      `
      : ''
    }
  }

  ${Rectangle} {
    border-radius: ${theme.$.radius.form.checkbox[shape]};
  }

  ${isChecked
    ? `
      ${Rectangle} {
        background: ${!isSoft ? theme.$.palette[palette].main : theme.$.palette[palette].soft};
      }

      ${Svg} > path {
        stroke-dashoffset: 0;
      }
    `
    : ''
  }
`)
