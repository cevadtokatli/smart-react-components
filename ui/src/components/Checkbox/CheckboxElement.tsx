import Div from '@smart-react-components/core/Element/Div'
import { StyledProps } from '@smart-react-components/core/styled-props'
import { PaletteProp, Partial, ResponsiveProp, ShapeProp, SizeProp } from '@smart-react-components/core/types'
import React from 'react'
import styled from 'styled-components'

const Container = styled.div.attrs({
  className: 'src-checkbox-container',
})(({ theme }) => `
  border: solid 2px ${theme.$.color.dynamic.accent};
  box-sizing: border-box;
  display: flex;
  height: 100%;
  justify-content: center;
  position: relative;
`)

const Rectangle = styled.div.attrs({
  className: 'src-checkbox-rectangle',
})`
  position: absolute;
  left: -2px;
  height: calc(100% + 4px);
  top: -2px;
  transition: background 200ms 0s ease-in-out;
  width: calc(100% + 4px);
`

const Svg = styled.svg.attrs({
  children: <path d="M3.5,17.2l7.6,7.6L28.5,7.3" />,
  className: 'src-checkbox-svg',
  viewBox: '0 0 32 32',
})(({ theme }) => `
  flex: 0 0 auto;
  position: relative;

  > path {
    fill: transparent;
    stroke: ${theme.$.color.dynamic.background};
    stroke-width: 4px;
    stroke-dasharray: 36px;
    stroke-dashoffset: 36px;
    transition: stroke-dashoffset 200ms 0s ease-in-out;
  }
`)

const IndeterminateMark = styled.div.attrs({
  className: 'src-checkbox-indeterminate-mark',
})(({ theme }) => `
  background: ${theme.$.color.dynamic.background};
  height: 2px;
  margin-top: -1px;
  opacity: 0;
  position: absolute;
  trasnform: scaleX(0) rotate(45deg);
  transition-duration: 200ms;
  transition-property: opacity, transform;
  transition-timing-function: ease-in-out;
  top: 50%;
  width: 100%;
`)

export interface Props extends
  StyledProps,
  Partial<ResponsiveProp<'checkboxSize', SizeProp>> {
  isChecked?: boolean
  isIndeterminate?: boolean
  isOutline?: boolean
  isSoft?: boolean
  palette?: PaletteProp
  shape?: ShapeProp
}

export default styled(Div).attrs<Props>(({ children, isIndeterminate }) => ({
  children: (
    <>
      { children && children }
      <Container>
        <Rectangle />
        <Svg />
        { typeof isIndeterminate !== 'undefined' && <IndeterminateMark /> }
      </Container>
    </>
  ),
  getCheckboxSize: (v, t) => `
    height: ${t.$.size.checkbox[v]};
    width: ${t.$.size.checkbox[v]};
  `,
}))<Props>(({ theme, isChecked, isIndeterminate, isOutline, isSoft, palette, shape }: Props) => `
  position: relative;

  .src-checkbox-container {
    border-radius: ${theme.$.radius.checkbox[shape]};
    transition: border-color 200ms 0s ease-in-out;

    ${!isOutline
      ? `
        background: ${theme.$.color.dynamic.accent};
      `
      : ''
    }
  }

  .src-checkbox-rectangle {
    border-radius: ${theme.$.radius.checkbox[shape]};
  }

  ${isChecked
    ? `
      .src-checkbox-container {
        border-color: ${!isSoft ? theme.$.palette[palette].main : theme.$.palette[palette].soft};

        .src-checkbox-rectangle {
          background: ${!isSoft ? theme.$.palette[palette].main : theme.$.palette[palette].soft};
        }

        ${!isIndeterminate
          ? `
            .src-checkbox-svg > path {
              stroke-dashoffset: 0;
            }
          `
          : `
            .src-checkbox-indeterminate-mark {
              opacity: 1;
              transform: scaleX(1);
            }
          `
        }
      }
    `
    : ''
  }
`)
