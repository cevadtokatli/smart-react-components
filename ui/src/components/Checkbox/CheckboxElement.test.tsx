import { wrapTheme } from '@smart-react-components/core/test'
import { render } from '@testing-library/react'
import React from 'react'
import CheckboxElement from './CheckboxElement'

describe('<CheckboxElement />', () => {
  let props

  beforeAll(() => {
    props = {
      checkboxSize: 'medium',
      isChecked: true,
      isOutline: false,
      isSoft: false,
      palette: 'primary',
      shape: 'rectangle',
    }
  })

  it('should render component', () => {
    const { asFragment } = render(wrapTheme(<CheckboxElement {...props} />))
    expect(asFragment()).toMatchSnapshot()
  })

  it('should not be checked', () => {
    const { asFragment } = render(wrapTheme(<CheckboxElement {...props} isChecked={false} />))
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render component outline', () => {
    const { asFragment } = render(wrapTheme(<CheckboxElement {...props} isOutline />))
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render component with soft colors', () => {
    const { asFragment } = render(wrapTheme(<CheckboxElement {...props} isSoft />))
    expect(asFragment()).toMatchSnapshot()
  })
})
