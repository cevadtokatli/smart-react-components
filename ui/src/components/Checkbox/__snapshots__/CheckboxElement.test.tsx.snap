// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxElement /> should not be checked 1`] = `
<DocumentFragment>
  .c0 {
  height: 16px;
  width: 16px;
}

.c3 {
  border: solid 2px #dbdbdb;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
}

.c5 {
  position: absolute;
  left: -2px;
  height: calc(100% + 4px);
  top: -2px;
  -webkit-transition: background 200ms 0s ease-in-out;
  transition: background 200ms 0s ease-in-out;
  width: calc(100% + 4px);
}

.c6 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  position: relative;
}

.c6 > path {
  fill: transparent;
  stroke: #fff;
  stroke-width: 4px;
  stroke-dasharray: 36px;
  stroke-dashoffset: 36px;
  -webkit-transition: stroke-dashoffset 200ms 0s ease-in-out;
  transition: stroke-dashoffset 200ms 0s ease-in-out;
}

.c1 {
  position: relative;
}

.c1 .c2 {
  border-radius: .125rem;
  background: #dbdbdb;
}

.c1 .c4 {
  border-radius: .125rem;
}

<div
    class="c0 c1"
    shape="rectangle"
  >
    <div
      class="c2 c3"
    >
      <div
        class="c4 c5"
      />
      <svg
        class="c6"
        viewBox="0 0 32 32"
      >
        <path
          d="M3.5,17.2l7.6,7.6L28.5,7.3"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<CheckboxElement /> should render component 1`] = `
<DocumentFragment>
  .c0 {
  height: 16px;
  width: 16px;
}

.c3 {
  border: solid 2px #dbdbdb;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
}

.c5 {
  position: absolute;
  left: -2px;
  height: calc(100% + 4px);
  top: -2px;
  -webkit-transition: background 200ms 0s ease-in-out;
  transition: background 200ms 0s ease-in-out;
  width: calc(100% + 4px);
}

.c7 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  position: relative;
}

.c7 > path {
  fill: transparent;
  stroke: #fff;
  stroke-width: 4px;
  stroke-dasharray: 36px;
  stroke-dashoffset: 36px;
  -webkit-transition: stroke-dashoffset 200ms 0s ease-in-out;
  transition: stroke-dashoffset 200ms 0s ease-in-out;
}

.c1 {
  position: relative;
}

.c1 .c2 {
  border-radius: .125rem;
  background: #dbdbdb;
}

.c1 .c4 {
  border-radius: .125rem;
}

.c1 .c4 {
  background: #0665d0;
}

.c1 .c6 > path {
  stroke-dashoffset: 0;
}

<div
    class="c0 c1"
    shape="rectangle"
  >
    <div
      class="c2 c3"
    >
      <div
        class="c4 c5"
      />
      <svg
        class="c6 c7"
        viewBox="0 0 32 32"
      >
        <path
          d="M3.5,17.2l7.6,7.6L28.5,7.3"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<CheckboxElement /> should render component outline 1`] = `
<DocumentFragment>
  .c0 {
  height: 16px;
  width: 16px;
}

.c3 {
  border: solid 2px #dbdbdb;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
}

.c5 {
  position: absolute;
  left: -2px;
  height: calc(100% + 4px);
  top: -2px;
  -webkit-transition: background 200ms 0s ease-in-out;
  transition: background 200ms 0s ease-in-out;
  width: calc(100% + 4px);
}

.c7 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  position: relative;
}

.c7 > path {
  fill: transparent;
  stroke: #fff;
  stroke-width: 4px;
  stroke-dasharray: 36px;
  stroke-dashoffset: 36px;
  -webkit-transition: stroke-dashoffset 200ms 0s ease-in-out;
  transition: stroke-dashoffset 200ms 0s ease-in-out;
}

.c1 {
  position: relative;
}

.c1 .c2 {
  border-radius: .125rem;
}

.c1 .c4 {
  border-radius: .125rem;
}

.c1 .c4 {
  background: #0665d0;
}

.c1 .c6 > path {
  stroke-dashoffset: 0;
}

<div
    class="c0 c1"
    shape="rectangle"
  >
    <div
      class="c2 c3"
    >
      <div
        class="c4 c5"
      />
      <svg
        class="c6 c7"
        viewBox="0 0 32 32"
      >
        <path
          d="M3.5,17.2l7.6,7.6L28.5,7.3"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<CheckboxElement /> should render component with soft colors 1`] = `
<DocumentFragment>
  .c0 {
  height: 16px;
  width: 16px;
}

.c3 {
  border: solid 2px #dbdbdb;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
}

.c5 {
  position: absolute;
  left: -2px;
  height: calc(100% + 4px);
  top: -2px;
  -webkit-transition: background 200ms 0s ease-in-out;
  transition: background 200ms 0s ease-in-out;
  width: calc(100% + 4px);
}

.c7 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  position: relative;
}

.c7 > path {
  fill: transparent;
  stroke: #fff;
  stroke-width: 4px;
  stroke-dasharray: 36px;
  stroke-dashoffset: 36px;
  -webkit-transition: stroke-dashoffset 200ms 0s ease-in-out;
  transition: stroke-dashoffset 200ms 0s ease-in-out;
}

.c1 {
  position: relative;
}

.c1 .c2 {
  border-radius: .125rem;
  background: #dbdbdb;
}

.c1 .c4 {
  border-radius: .125rem;
}

.c1 .c4 {
  background: #cde0f5;
}

.c1 .c6 > path {
  stroke-dashoffset: 0;
}

<div
    class="c0 c1"
    shape="rectangle"
  >
    <div
      class="c2 c3"
    >
      <div
        class="c4 c5"
      />
      <svg
        class="c6 c7"
        viewBox="0 0 32 32"
      >
        <path
          d="M3.5,17.2l7.6,7.6L28.5,7.3"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
